File: semgus-pretty/src/main/java/org/semgus/pretty/Code.java
package org.semgus.pretty;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Stream;

/**
 * Pretty code.
 * Learned from org.aya.pretty.
 */
public sealed interface Code extends Prettiable {
  @Override default @NotNull Code codify() {
    return this;
  }

  record Empty() implements Code {}

  record Char(char c) implements Code {}

  record Text(@NotNull String s) implements Code {}

  record Line() implements Code {}

  record FlatAlt(@NotNull Code l, @NotNull Code r) implements Code {}

  record Cat(@NotNull List<Code> xs) implements Code {}

  record Nest(int i, @NotNull Code x) implements Code {}

  // TODO: Not needed for now.
  record Union(@NotNull Code l, @NotNull Code r) implements Code {}

  static @NotNull Code empty() {
    return new Empty();
  }

  static @NotNull Code plain(char c) {
    return new Char(c);
  }

  static @NotNull Code plain(@NotNull String s) {
    return new Text(s);
  }

  static @NotNull Code nest(int i, @NotNull Code x) {
    return i == 0 ? x : new Nest(i, x);
  }

  static @NotNull Code flatAlt(@NotNull Code l, @NotNull Code r) {
    return new FlatAlt(l, r);
  }

  static @NotNull Code hardline() {
    return new Line();
  }

  static @NotNull Code line() {
    return flatAlt(hardline(), plain(' '));
  }

  static @NotNull Code line0() {
    return flatAlt(hardline(), empty());
  }

  // TODO: Not needed for now.
  static @NotNull Code group(@NotNull Code x) {
    return x instanceof FlatAlt xFlatAlt ? xFlatAlt.l : x;
  }

  static @NotNull Code encloseDelim(@NotNull Code l, @NotNull Code r, @NotNull Code d, @NotNull List<Code> xs) {
    return enclose(l, r, concatWith(d, xs));
  }

  static @NotNull Code list(@NotNull List<Code> xs) {
    return group(
      encloseDelim(
        plain('['),
        plain(']'),
        plain(", "),
        xs
      )
    );
  }

  static @NotNull Code list(@NotNull Code... xs) {
    return list(List.of(xs));
  }

  static @NotNull Code tupled(@NotNull List<Code> xs) {
    return group(
      encloseDelim(
        plain('('),
        plain(')'),
        plain(", "),
        xs
      )
    );
  }

  static @NotNull Code tupled(@NotNull Code... xs) {
    return tupled(List.of(xs));
  }

  static @NotNull Code concatWith(@NotNull Code d, @NotNull List<Code> xs) {
    return hcat(
      xs.stream()
        .flatMap(x -> Stream.of(d, x))
        .skip(1)
        .toList()
    );
  }

  static @NotNull Code hsep(@NotNull List<Code> xs) {
    return concatWith(plain(' '), xs);
  }

  static @NotNull Code hsep(@NotNull Code... xs) {
    return hsep(List.of(xs));
  }

  static @NotNull Code vsep(@NotNull List<Code> xs) {
    return concatWith(line(), xs);
  }

  static @NotNull Code vsep(@NotNull Code... xs) {
    return vsep(List.of(xs));
  }

  static @NotNull Code sep(@NotNull List<Code> xs) {
    return group(vsep(xs));
  }

  static @NotNull Code sep(@NotNull Code... xs) {
    return sep(List.of(xs));
  }

  static @NotNull Code hcat(@NotNull List<Code> xs) {
    return new Cat(xs);
  }

  static @NotNull Code hcat(@NotNull Code... xs) {
    return hcat(List.of(xs));
  }

  static @NotNull Code vcat(@NotNull List<Code> xs) {
    return concatWith(line0(), xs);
  }

  static @NotNull Code vcat(@NotNull Code... xs) {
    return vcat(List.of(xs));
  }

  static @NotNull Code cat(@NotNull List<Code> xs) {
    return group(vcat(xs));
  }

  static @NotNull Code cat(@NotNull Code... xs) {
    return cat(List.of(xs));
  }

  static @NotNull Code spaces(int n) {
    return n <= 0 ? empty()
      : n == 1 ? plain(' ')
      : plain(" ".repeat(n));
  }

  static @NotNull Code enclose(@NotNull Code l, @NotNull Code r, @NotNull Code x) {
    return hcat(l, x, r);
  }

  static @NotNull Code surround(@NotNull Code x, @NotNull Code l, @NotNull Code r) {
    return hcat(l, x, r);
  }

  static @NotNull Code squotes(@NotNull Code x) {
    return enclose(plain('\''), plain('\''), x);
  }

  static @NotNull Code dquotes(@NotNull Code x) {
    return enclose(plain('\"'), plain('\"'), x);
  }

  static @NotNull Code parens(@NotNull Code x) {
    return enclose(plain('('), plain(')'), x);
  }

  static @NotNull Code angles(@NotNull Code x) {
    return enclose(plain('<'), plain('>'), x);
  }

  static @NotNull Code brackets(@NotNull Code x) {
    return enclose(plain('['), plain(']'), x);
  }

  static @NotNull Code braces(@NotNull Code x) {
    return enclose(plain('{'), plain('}'), x);
  }

  static @NotNull Code verts(@NotNull Code x) {
    return enclose(plain('|'), plain('|'), x);
  }

  static @NotNull Code semi(@NotNull Code x) {
    return hcat(x, plain(';'));
  }

  static @NotNull Code block(@NotNull Code prefix, int i, @NotNull Code body) {
    return hcat(
      hsep(prefix, plain('{')),
      nest(i, hcat(
        line(),
        body
      )),
      line(),
      plain('}')
    );
  }
}

File: semgus-pretty/src/main/kotlin/Code.kt
package org.semgus.pretty

/**
 * Pretty code.
 * Learned from org.aya.pretty.
 */
sealed class Code : Prettiable {
  object Empty : Code()
  object Line : Code()
  data class CPlain(val c: Char) : Code()
  data class SPlain(val s: String) : Code()
  data class FlatAlt(val l: Code, val r: Code) : Code()
  data class Cat(val xs: Sequence<Code>) : Code()
  data class Nest(val i: Int, val x: Code) : Code()

  // TODO: Not needed for now.
  data class Union(val l: Code, val r: Code) : Code()

  override fun codify() = this
}

fun empty() = Code.Empty

fun space() = plain(' ')
fun spaces(n: Int) =
  if (n <= 0) empty()
  else if (n == 1) space()
  else plain(" ".repeat(n))

fun hardline() = Code.Line
fun line0() = flatAlt(hardline(), empty())
fun line() = flatAlt(hardline(), space())

fun plain(c: Char) = Code.CPlain(c)
fun plain(s: String) = Code.SPlain(s)

fun flatAlt(l: Code, r: Code) = Code.FlatAlt(l, r)

fun hcat(xs: Sequence<Code>) = Code.Cat(xs)
fun hcat(vararg xs: Code) = hcat(xs.asSequence())
fun vcat(xs: Sequence<Code>) = hcatDelim(line0(), xs)
fun vcat(vararg xs: Code) = vcat(xs.asSequence())
fun cat(xs: Sequence<Code>) = group(vcat(xs))
fun cat(vararg xs: Code) = cat(xs.asSequence())

fun hsep(xs: Sequence<Code>) = hcatDelim(space(), xs)
fun hsep(vararg xs: Code) = hsep(xs.asSequence())
fun vsep(xs: Sequence<Code>) = hcatDelim(line(), xs)
fun vsep(vararg xs: Code) = vsep(xs.asSequence())
fun sep(xs: Sequence<Code>) = group(vsep(xs))
fun sep(vararg xs: Code) = sep(xs.asSequence())

fun nest(i: Int, x: Code) =
  if (i == 0) x
  else Code.Nest(i, x)

// TODO: Not needed for now.
fun union(l: Code, r: Code) = Code.Union(l, r)

// TODO: Not needed for now.
fun group(x: Code) =
  if (x is Code.FlatAlt) x.l
  else x

fun hcatDelim(d: Code, xs: Sequence<Code>) = hcat(
  xs.flatMap { sequenceOf(d, it) }.drop(1),
)

fun hsepDelim(d: Code, xs: Sequence<Code>) = hsep(
  xs.flatMap { sequenceOf(d, it) }.drop(1),
)

fun surround(x: Code, l: Code, r: Code) = hcat(l, x, r)
fun enclose(l: Code, r: Code, x: Code) = hcat(l, x, r)

fun hcatEncloseDelim(l: Code, r: Code, d: Code, xs: Sequence<Code>) =
  enclose(l, r, hcatDelim(d, xs))

fun hsepEncloseDelim(l: Code, r: Code, d: Code, xs: Sequence<Code>) =
  enclose(l, r, hsepDelim(d, xs))

fun list(xs: Sequence<Code>) = group(
  hcatEncloseDelim(
    flatAlt(plain("[ "), plain('[')),
    flatAlt(plain(" ]"), plain(']')),
    plain(", "),
    xs,
  ),
)

fun list0(xs: Sequence<Code>) = hcatEncloseDelim(
  plain('['),
  plain(']'),
  plain(", "),
  xs,
)

fun tuple(xs: Sequence<Code>) = group(
  hcatEncloseDelim(
    flatAlt(plain("( "), plain('(')),
    flatAlt(plain(" )"), plain(')')),
    plain(", "),
    xs,
  ),
)

fun tuple0(xs: Sequence<Code>) = hcatEncloseDelim(
  plain('('),
  plain(')'),
  plain(", "),
  xs,
)

fun set(xs: Sequence<Code>) = group(
  hcatEncloseDelim(
    flatAlt(plain("{ "), plain('{')),
    flatAlt(plain(" }"), plain('}')),
    plain(", "),
    xs,
  ),
)

fun set0(xs: Sequence<Code>) = hcatEncloseDelim(
  plain('{'),
  plain('}'),
  plain(", "),
  xs,
)

fun block(i: Int, prefix: Code, body: Code) = hcat(
  hsep(prefix, plain('{')),
  nest(i, hcat(line(), body)),
  line(),
  plain('}'),
)

fun squotes(x: Code) = enclose(plain('\''), plain('\''), x)
fun dquotes(x: Code) = enclose(plain('\"'), plain('\"'), x)
fun parens(x: Code) = enclose(plain('('), plain(')'), x)
fun angles(x: Code) = enclose(plain('<'), plain('>'), x)
fun brackets(x: Code) = enclose(plain('['), plain(']'), x)
fun braces(x: Code) = enclose(plain('{'), plain('}'), x)
fun verts(x: Code) = enclose(plain('|'), plain('|'), x)
fun semi(x: Code) = hcat(x, plain(';'))
